### PR TITLE
Ignore removed and renamed files when linting

### PR DESCRIPTION
**Pull request by xiangpengzhao**
fixes: #4599

/cc @kargakis @spxtr

### CODE REVIEW

**Comment by kargakis** [comment]
Nice

Can you add a unit test for `modifiedGoFiles`?

**Comment by xiangpengzhao** [comment]
@kargakis ok, will add it when I'm back to my keyboard tomorrow.

**Comment by spxtr** [comment]
Awesome, thanks. Can you also verify that it works for renamed files?

File: prow/plugins/golint/golint.go [hunk 1]
```diff
@@ -73,11 +73,14 @@ func modifiedGoFiles(ghc githubClient, org, repo string, number int, sha string)
...
+		switch {
+		case strings.HasPrefix(change.Filename, "vendor/"):
 			continue
-		} else if filepath.Ext(change.Filename) != ".go" {
+		case filepath.Ext(change.Filename) != ".go":
 			continue
-		} else if gfg.Match(change.Filename) {
+		case gfg.Match(change.Filename):
+			continue
+		case change.Status == github.PullRequestFileRemoved:
```
Comment (xiangpengzhao) [inline prow/plugins/golint/golint.go]: The `PullRequestFileRenamed` file can also be ignored.

File: prow/github/types.go [hunk 2]
```diff
@@ -122,11 +122,21 @@ type Label struct {
 	Color string `json:"color"`
 }
 
+// PullRequestFileStatus enumerates the statuses for this webhook payload type.
+type PullRequestFileStatus string
+
+const (
+	PullRequestFileModified PullRequestFileStatus = "modified"
```
Comment (xiangpengzhao) [inline prow/github/types.go]: I didn't find statuses on github API pages https://developer.github.com/v3/.

Just found something here https://github.com/github/VisualStudio/blob/5baf8379abdc0356dffa356f2d66855aaedf063f/src/GitHub.Exports/Models/IPullRequestFileModel.cs

```
    public enum PullRequestFileStatus
    {
        Modified,
        Added,
        Removed,
        Renamed,
    }
```

I think this is what we want.

**Comment by xiangpengzhao** [comment]
>Can you also verify that it works for renamed files?

@spxtr Just see your comment after I comment inline and refresh the page :) I think the lint can ignore renamed files. Will add test case to verify this.

**Comment by spxtr** [comment]
> I think the lint can ignore renamed files. Will add test case to verify this.

I don't think it needs to. It seems pretty likely that it will just work, but I want to make sure.

**Comment by xiangpengzhao** [comment]
I'm not sure how to make sure:)

**Comment by xiangpengzhao** [comment]
Ah, got it. I was thinking you meant verifying this before we merge/deploy this PR :)

**Comment by spxtr** [comment]
LGTM once a unit test is in place.

**Comment by xiangpengzhao** [comment]
Will add it once I'm on keyboard. (on smartphone now)

**Comment by xiangpengzhao** [comment]
Test case added. @kargakis @spxtr PTAL. Thanks!

File: prow/plugins/golint/golint_test.go [hunk 3]
```diff
@@ -240,3 +241,136 @@ func TestAddedLines(t *testing.T) {
...
+		{
+			name: "modified files include non go file",
+			gh: &ghc{
+				changes: []github.PullRequestChange{
+					{
+						Filename: "qux.go",
+						Patch:    "@@ -0,0 +1,5 @@\n+package bar\n+\n+func Qux() error {\n+   return nil\n+}",
+					},
+					{
+						Filename: "foo.come",
```
Comment (kargakis) [inline prow/plugins/golint/golint_test.go]: Use a common filename such as BUILD, or .md, or .yaml.

File: prow/plugins/golint/golint_test.go [hunk 4]
```diff
@@ -240,3 +241,136 @@ func TestAddedLines(t *testing.T) {
...
+						Status:   github.PullRequestFileRenamed,
+						Patch:    "@@ -0,0 +1,5 @@\n+package bar\n+\n+func Qux2() error {\n+   return nil\n+}",
+					},
+				},
+			},
+			expectedModifiedFiles: map[string]string{
+				"qux.go": "@@ -0,0 +1,5 @@\n+package bar\n+\n+func Qux() error {\n+   return nil\n+}",
+			},
+		},
+	}
```
Comment (kargakis) [inline prow/plugins/golint/golint_test.go]: Add two more test cases: one where more than a single modified file is expected, and another one where none of the changes are expected.

**Comment by xiangpengzhao** [comment]
@kargakis test case updated. PTAL. Thanks!

**Comment by xiangpengzhao** [comment]
hook was not updated in #4627, this PR should be deployed with that one.

**Comment by spxtr** [comment]
Thanks for the quick fix on this one and #4627!

**Comment by xiangpengzhao** [comment]
My pleasure! I like the cute bots and hope they can be smarter :smile:

**Comment by xiangpengzhao** [comment]
`/lint` even not works for a modified file. See in: https://github.com/kubernetes/kubernetes/pull/52888#issuecomment-331369344 I believe the modified file has a lint error.

**Comment by xiangpengzhao** [comment]
Maybe it's because this PR hasn't been deployed yet?

**Comment by xiangpengzhao** [comment]
Verified in https://github.com/kubernetes/kubernetes/pull/52888.

**Comment by spxtr** [comment]
Yay! Good stuff.

**Pull request merged**